typed_syscall_data/README.md

```diff
@@ -0,0 +1,481 @@
+types](https://doc.rust-lang.org/reference/types.html) and see several types
+that probably make sense in a syscall interface:
+
+* `bool`
+* Signed integers: `i8`, `i16`, `i32`, `i64`, `i128`, `isize`
+* Unsigned integers: `u8`, `u16`, `u32`, `u64`, `u128`, `usize`
+* Floating point: `f32`, `f64`
+* Codepoint: `char`
+* Upcall function pointer
+* Raw pointers: `*const T`, `*mut T`
+
+There are a couple other types that we probably want to consider as well:
+
+* Tock's `ErrorCode`, as it is extremely common.
+* A non-pointer CHERI capability.
+* A Register type that represents *any* possible register value. This is a
+  future-compatibility safeguard: if we ever need a type that is not in our
+  fixed list, we can call it a Register and make things work (albeit with less
+  type safety). It also allows us to embed ArbitraryData (which will be defined
+  later) inside ArbitraryData, which is used by Yield.
+
+### Reducing the number of types
+
+Of course, there is a cost to complexity and in particular supporting a long
+list of types. It probably doesn't make sense to provide special support for
+*all* of the above types. In particular:
+
+* `i8` and `i16` are not particularly common and can easily be passed as `i32`.
+* `u8` and `u16` can similarly be passed as `u32`.
+* `i128` and `u128` are extremely rare in Rust, completely absent in standard C,
+  and probably exceptionally rare on the small systems Tock targets.
+* `char` is rare in Tock, as Tock mostly treats text as byte buffers. When
+  needed, it can be passed as `u32`. It's also not a standard part of C (where
+  `char` has a different meaning).
+* Wide pointers: these aren't a concept in other languages (e.g. C) and
+  therefore don't belong in a syscall ABI.
+* `*const T` and `*mut T` are redundant: they're distinct types, but only exist
+  to communicate mutability information. We already handle this in Tock by using
+  different system calls for buffers of different mutability. Therefore we only
+  need one pointer type.
+
+### The reduced list
+
+This leaves us with:
+
+* `bool`
+* Numeric: `i32`, `i64`, `isize`, `u32`, `u64`, `usize`, `f32`, `f64`
+* Upcall function pointer
+* Narrow pointer (`*const T` or `*mut T` where `T: Sized`)
+* `ErrorCode`
+* Non-pointer CHERI capability
+
+## Type descriptors
+
+When a process wants to send data to the kernel, the process needs a way to tell
+the kernel the sent data's type (and vice versa when the kernel sends data to
+the process). To do this, we need a way to serialize information about a a list
+of types. To start, lets assign numbers to each type (DNE means this type
+doesn't exist yet):
+
+| ID       | Description                  | Kernel Type     | `libtock-c` Type | `libtock-rs` Type         |
+| -------- | ---------------------------- | --------------- | ---------------- | ------------------------- |
+| `0b0001` | Error code                   | `ErrorCode`     | DNE              | `ErrorCode`               |
```

Comment:
Add `0b0000` to the table as "invalid" or something like that.

Reply:
Done.